model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java

```diff
@@ -0,0 +1,140 @@
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.IdentityHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * {@link MapKeycloakTransaction} implementation used with the file map storage.
+ *
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+public class FileMapKeycloakTransaction<V extends AbstractEntity & UpdatableEntity, M>
+  extends ConcurrentHashMapKeycloakTransaction<String, V, M> {
+
+    private final List<Path> touchedPaths = new LinkedList<>();
+    private Map<Path, Path> renameOnCommit = new IdentityHashMap<>();
+
+    private final String txId = StringKey.INSTANCE.yieldNewUniqueKey();
+
+    public static <V extends AbstractEntity & UpdatableEntity, M> FileMapKeycloakTransaction<V, M> newInstance(Class<V> entityClass,
+      Function<String, Path> dataDirectoryFunc, Function<V, String[]> suggestedPath,
+      boolean isExpirableEntity, Map<SearchableModelField<? super M>, UpdatePredicatesFunc<String, V, M>> fieldPredicates) {
+        Crud<V, M> crud = new Crud<>(entityClass, dataDirectoryFunc, suggestedPath, isExpirableEntity, fieldPredicates);
+        FileMapKeycloakTransaction<V, M> tx = new FileMapKeycloakTransaction<>(entityClass, crud);
+        crud.tx = tx;
+        return tx;
+    }
+
+    private FileMapKeycloakTransaction(Class<V> entityClass, Crud<V, M> crud) {
+        super(
+          crud,
+          StringKeyConverter.StringKey.INSTANCE,
+          DeepCloner.DUMB_CLONER,
+          MapFieldPredicates.getPredicates(ModelEntityUtil.getModelType(entityClass)),
+          ModelEntityUtil.getRealmIdField(entityClass)
+        );
+    }
+
+    @Override
+    public void rollback() {
+        this.touchedPaths.forEach(FileMapKeycloakTransaction::silentDelete);
+        super.rollback();
+    }
+
+    @Override
+    public void commit() {
+        super.commit();
+        this.renameOnCommit.forEach(FileMapKeycloakTransaction::silentMove);
+        this.touchedPaths.forEach(FileMapKeycloakTransaction::silentDelete);
+    }
+
+    private static void silentMove(Path from, Path to) {
+        try {
+            Files.move(from, to, StandardCopyOption.REPLACE_EXISTING);
+        } catch (IOException ex) {
+            throw new UncheckedIOException(ex);
+        }
+    }
+
+    private static void silentDelete(Path p) {
+        try {
+            if (Files.exists(p) && Files.size(p) == 0) {
+                Files.delete(p);
+            }
+        } catch (IOException e) {
+            // Swallow the exception
+        }
+    }
+
+    public void touch(Path path) throws FileAlreadyExistsException, IOException {
+        Files.createFile(path);
+        touchedPaths.add(path);
+    }
+
+    void registerRenameOnCommit(Path from, Path to) {
+        this.renameOnCommit.put(from, to);
+        this.touchedPaths.add(from);
```

Comment:
When debugging this, I was surprised that the `from` path was added to the `touchedPaths`. 

As the `renameOnCommit` is executed first on the happy commit path, the removing of all touched paths doesn't do anything on the `fromfile` - it is already moved to the new name.

On the unhappy path on rollback, that `from` file will only be removed when it is still empty. It's a bit hard for me to follow on this logic to see if this leads always to the correct results, and if the deletion of files on rollback could be determined in a simpler way.

Reply:
`from` files are temporary files owned by current transaction, e.g. `.tx-123-realm.yaml`. They have to be removed on commit (via renaming: `silentMove`) or on rollback (deleting).